index.js

```diff
@@ -182,216 +144,68 @@
-   *
-   * @private
-   */
-  FastDom.prototype.runBatch = function() {
     try {
-
-      // Set the mode to 'reading',
-      // then empty all read jobs
-      this.batch.mode = 'reading';
-      this.flush(this.batch.read);
-
-      // Set the mode to 'writing'
-      // then empty all write jobs
-      this.batch.mode = 'writing';
-      this.flush(this.batch.write);
-
-      this.batch.mode = null;
-
+      if (this.runBatch(this.reads, 10000, start, readBatchTimeLimit)) {
+        if (this.runBatch(this.writes, 10000, start, writeBatchTimeLimit)) {
+          this.runBatch(this.deferred, this.deferred.length, start, deferredTimeLimit); // deferred.length ensures newly deferred jobs aren't run immediately
+        }
+      }
     } catch (e) {
-      this.runBatch();
-      throw e;
+      error = e;
     }
-  };
-
-  /**
-   * Adds a new job to
-   * the given batch.
-   *
-   * @param {Array}   list
-   * @param {Function} fn
-   * @param {Object}   ctx
-   * @returns {Number} id
-   * @private
-   */
-  FastDom.prototype.add = function(type, fn, ctx) {
-    var id = this.uniqueId();
-    return this.batch.hash[id] = {
-      id: id,
-      fn: fn,
-      ctx: ctx,
-      type: type
-    };
-  };
-
-  /**
-   * Runs a given job.
-   *
-   * Applications using FastDom
-   * have the options of setting
-   * `fastdom.onError`.
-   *
-   * This will catch any
-   * errors that may throw
-   * inside callbacks, which
-   * is useful as often DOM
-   * nodes have been removed
-   * since a job was scheduled.
-   *
-   * Example:
-   *
-   *   fastdom.onError = function(e) {
-   *     // Runs when jobs error
-   *   };
-   *
-   * @param  {Object} job
-   * @private
-   */
-  FastDom.prototype.run = function(job){
-    var ctx = job.ctx || this;
-    var fn = job.fn;
-
-    // Clear reference to the job
-    delete this.batch.hash[job.id];
 
-    // If no `onError` handler
-    // has been registered, just
-    // run the job normally.
-    if (!this.onError) {
-      return fn.call(ctx);
+    this.scheduled = false;
+    if (this.reads.length || this.writes.length || this.deferred.length) {
+      this.scheduleBatch();
     }
 
-    // If an `onError` handler
-    // has been registered, catch
-    // errors that throw inside
-    // callbacks, and run the
-    // handler instead.
-    try { fn.call(ctx); } catch (e) {
-      this.onError(e);
-    }
-  };
-
-  /**
-   * Starts a rAF loop
-   * to empty the frame queue.
-   *
-   * @private
-   */
-  FastDom.prototype.loop = function() {
-    var self = this;
-    var raf = this.raf;
-
-    // Don't start more than one loop
-    if (this.looping) return;
-
-    raf(function frame() {
-      var fn = self.frames.shift();
-
-      // If no more frames,
-      // stop looping
-      if (!self.frames.length) {
-        self.looping = false;
-
-      // Otherwise, schedule the
-      // next frame
+    if (error) {
+      if (this.onError) {
+        this.onError(error);
```

Comment:
I'm tempted to just get rid of `.onError` altogether. We should just throw all the time.